paimon-core/src/main/java/org/apache/paimon/index/PartitionIndex.java

```diff
@@ -80,29 +86,35 @@ public int assign(int hash, IntPredicate bucketFilter) {
             Long number = entry.getValue();
             if (number < targetBucketRowNumber) {
                 entry.setValue(number + 1);
-                hash2Bucket.put(hash, bucket.shortValue());
-                return bucket;
+                return cacheBucketAndGet(hash2Bucket, hash, bucket);
             } else {
                 iterator.remove();
             }
         }
 
-        // 3. create a new bucket
-        for (int i = 0; i < Short.MAX_VALUE; i++) {
-            if (bucketFilter.test(i) && !totalBucket.contains(i)) {
-                hash2Bucket.put(hash, (short) i);
-                nonFullBucketInformation.put(i, 1L);
-                totalBucket.add(i);
-                return i;
+        if (-1 == maxBucketsNum || totalBucket.size() < maxBucketsNum) {
```

Comment:
I think here should be: max bucket should be smaller than maxBucketsNum.

If it is a job restarted from scratch, each task is increasing, and the previous judgment may be problematic.

Reply:
Wise consideration! done :)